src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs

```diff
@@ -44,153 +46,70 @@ private static ILiteralExpression CreateIncrementOneLiteralExpression(BoundIncre
-                              (argument.Type.TypeKind != TypeKind.Array ||
-                              !argument.Type.Equals((CSharp.Symbols.TypeSymbol)parameters[parameters.Length - 1].Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)))))
-                        {
-                            var parameter = parameters[parameters.Length - 1];
-                            var value = CreateParamArray(parameter, boundArguments, argumentIndex, invocationSyntax);
-
-                            return new Argument(
-                                argumentKind: ArgumentKind.ParamArray,
-                                parameter: parameter,
-                                value: value,
-                                inConversion: null,
-                                outConversion: null,
-                                isInvalid: parameter == null || value.IsInvalid,
-                                syntax: value.Syntax,
-                                type: null,
-                                constantValue: default(Optional<object>));
-                        }
-                        else
-                        {
-                            var value = Create(argument);
-                            return new Argument(
-                                argumentKind: ArgumentKind.Explicit,
-                                parameter: parameterOpt,
-                                value: value,
-                                inConversion: null,
-                                outConversion: null,
-                                isInvalid: parameterOpt == null || value.IsInvalid,
-                                syntax: value.Syntax,
-                                type: null,
-                                constantValue: default(Optional<object>));
-                        }
+                        // optionalParametersMethod can be null if we are writing to a readonly indexer or reading from an writeonly indexer,
+                        // in which case HasErrors property would be true, but we still want to treat this as invalid invocation.
+                        return boundArguments.SelectAsArray(arg => CreateArgumentOperation(ArgumentKind.Explicit, null, Create(arg)));
                     }
 
-                    var operation = Create(argument);
-                    return new Argument(
-                        argumentKind: ArgumentKind.Explicit,
-                        parameter: parameterOpt,
-                        value: operation,
-                        inConversion: null,
-                        outConversion: null,
-                        isInvalid: parameterOpt == null || operation.IsInvalid,
-                        syntax: operation.Syntax,
-                        type: null,
-                        constantValue: default(Optional<object>));
+                    return LocalRewriter.MakeArgumentsInEvaluationOrder(
```

Comment:
There's another method in local rewriter that does similar stuff during lowering that shares a lot of logic with this one, so it might be better to keep them together.